### DISCUSSION TITLE
Support always inlined methods

### DISCUSSION THREAD

**Discussion by AlexRadch** [dotnet/csharplang#7290] [Language Ideas] [answered]
When developing productive code, you now often have to work with structures located on the stack. They cannot be created and initialized in some helper method and then returned to a usual method.

You have to write all this high-performance shell in the body of a usual method. This makes it challenging to develop high-performance methods and destroys the purity and beauty of productive code...

For example, I can't write the following helper method and use it in a usual method:

```cs
// typical bytes initialization that can be used in many places
public static inline Span<byte> GetBytesOnStack(this Encoding encoding, string str) // inline keyword
{
    Span<byte> bytes = stackalloc byte[encoding.GetByteCount(str)];
    encoding.GetBytes(str, bytes);
    return bytes; // Cannot use variable 'bytes' in this context because it may expose referenced variables outside of their declaration scope
}

bool SomeOptimizedMethod(string str)
{
    Span<byte> bytes = Encoding.UTF8.GetBytesOnStack(str);
    // work with bytes
}

// It will be compiled to
bool SomeOptimizedMethod(string str)
{
    // Compiler inlined code
        Encoding _generated_name1 = Encoding.UTF8;
        Span<byte> _generated_name2 = stackalloc byte[_generated_name1.GetByteCount(str)];
        _generated_name1.GetBytes(str, _generated_name2);
        Span<byte> bytes = _generated_name2;
    // End compiler inlined code

    // work with bytes
}
```

I suggest adding a new `inline` keyword when defining such methods. The `inline` keyword will ensure that the body of a helper method will be inserted into the body of a method right at COMPILATION time, such methods will not be inserted by the JIT compiler, they will work like small macro commands at compile time and improve the writing of high-performance code. They allow returning beauty and the convenience to a high-performance code.

The suggestion can be extended to have helper methods generate both initialization code and finalization code... if needed... But that would be a more extended suggestion and can be discussed later if the inial suggestion is well received...

**Comment by HaloFour** [comment]
See: #1413

**Comment by AlexRadch** [comment]
F# is already support inlining see example for my code: https://sharplab.io/#v2:DYLgZgzgNAJiDUAfAsAKAPYAcCmA7ABALICWAxgE7oTpgAuAdAGIDKAFgIbmb0By7txAG7YAkrlrZKmNFjz5mATwgSAtvQAq2AB60ZOAouXYVaNMGy18xXMGvZ8AcwsAhBRIgB5XM1rtSAa3wACjxSdBhrBxB8AFFcMIjcBwBKYOVyaPTI1IBeNHwC/HNLcyTaVnwc/FDwyPoAcRc3bABhdABXcXx0/MLi/EF0YhgABVpySt7Cwr4BYTHyemU/f3ZgYHRSAB4AI2aAPiK8B3Kp6fxEQ9mhbAX6WnQANSHR8bPmTHZcXYOgweGFlAjmVWMlTKh+tQVNgPJgBCpiAAvbAwQgWVjhbrjSaoab9PbuSqOJruLw+FaxeK1JL0ACq6kYAA4seQzkEwaggA
```fs
let inline getBytesOnStack (encoding: Encoding) (str: string) =
    let length = encoding.GetByteCount str
    let voidPtr =
        NativePtr.stackalloc<byte> length
        |> NativePtr.toVoidPtr
    Span<byte>(voidPtr, length)

let someOptimizedMethod str =
    let bytes = getBytesOnStack Encoding.UTF8 str
    ()
```

will generate:
```cs
    public unsafe static void someOptimizedMethod(string str)
    {
        Encoding uTF = Encoding.UTF8;
        int byteCount = uTF.GetByteCount(str);
        Span<byte> span = new Span<byte>(stackalloc byte[byteCount], byteCount);
    }
```

**Comment by HaloFour** [comment]
Your example includes a `public` method that is to be inlined.  How do you suggest that a feature like this works across assembly boundaries when the original C# source code no longer exists?

**Reply by Szer** [reply]
F# inline methods work across assembly boundaries. All inline methods AST are written to assembly metadata and inlined at compile time

FSharp.Core (standard library implicitly included in F# projects) has a lot of inline methods, which are properly inlined in user assemblies

**Reply by HaloFour** [reply]
F# uses an enormous amount of custom metadata, and any AST embedded in the assembly would definitely be on that list.  Such an AST would be compiler specific, to the extent that such a feature could not be a language feature but would depend entirely on the inner workings of the specific compiler implementation.  The entire Roslyn AST would have to become a part of the C# specification.  IMO, that would make it a very ~poor~ challenging choice for a language feature.

**Reply by AlexRadch** [reply]
I think AST is IL AST and is source language-independent.

**Reply by HaloFour** [reply]
IL doesn't have an AST, just the raw metadata tables and bytecode.  The C# language also doesn't specifically compile to IL.

**Reply by Szer** [reply]
Nah, F# inline methods are written in F# AST, IL can't be written there, as inline methods of F# have a lot of compile-time type resolution features which can't be done with IL.
Also F# compiled not only to IL, but to JS, Rust, Python as well, and inline methods work in these scenarios as well

**Reply by AlexRadch** [reply]
Can .Net Expression classes be used here? They look like AST trees.

**Reply by siegfriedpammer** [reply]
Note that the expression tree API only supports a small subset of C# 3.0. There is still an ongoing discussion on how to expand support for newer language constructs. I don't think expression trees are the way to go.

**Comment by qrli** [comment, marked as answer]
Instead of a helper method, you can use a helper `ref struct` instead, which can be constructed on stack and also encapsulate the logic.

**Reply by AlexRadch** [reply]
Yes, you are right! I have already been shown this interesting trick and how the dotnet development team uses it. Thank you.

**Reply by huoyaoyuan** [reply]
But the actual stack allocation size can't be dynamically decided then. The pattern is quite common in BCL.
